src/firestore-adapter.ts

```diff
@@ -0,0 +1,859 @@
+    const ops: ITextOp[] = revision.operation.getOps().slice(0, 2);
+
+    let position = 0;
+
+    for (const op of ops) {
+      if (op.isRetain()) {
+        position += op.chars!;
+      }
+
+      if (op.isInsert()) {
+        position += op.text!.length;
+      }
+    }
+
+    setTimeout(() => {
+      if (this._zombie) {
+        return;
+      }
+
+      const { color, name } = this._userColorMap.get(author);
+
+      this._trigger(
+        FirebaseAdapterEvent.CursorChange,
+        author,
+        { position, selectionEnd: position },
+        color,
+        name
+      );
+    }, 1);
+
+    if (this._typingState.has(author)) {
+      clearTimeout(this._typingState.get(author));
+      this._typingState.delete(author);
+    }
+
+    const timerId = setTimeout(() => {
+      /**
+       * Unblock remote users for
+       * 1. Code range Highlight events
+       * 2. Code navigation using keyboard events
+       */
+      this._typingState.delete(author);
+    }, FirestoreAdapter.CURSOR_EVENT_DEFER_DELAY);
+
+    /**
+     * Block early arriving cursor events when remote
+     * user is typing
+     */
+    this._typingState.set(author, timerId);
+  }
+
+  sendOperation(
+    operation: TextOperation,
+    callback: SendOperationCallbackType = Utils.noop
+  ): void {
+    // If we're not ready yet, do nothing right now, and trigger a retry when we're ready.
+    if (!this._ready) {
+      this.on(FirebaseAdapterEvent.Ready, () => {
+        this._trigger(FirebaseAdapterEvent.Retry);
+      });
+      return;
+    }
+
+    // Sanity check that this operation is valid.
+    if (!this._document!.canMergeWith(operation)) {
+      const error = "sendOperation() called with invalid operation.";
+      this._trigger(FirebaseAdapterEvent.Error, error, operation.toString(), {
+        operation: operation.toString(),
+        document: this._document!.toString(),
+      });
+      Utils.onInvalidOperationRecieve(error);
+    }
+
+    // Convert revision into an id that will sort properly lexicographically.
+    const revisionId = this._revisionToId(this._revision);
+
+    this._sent = { id: revisionId, op: operation };
+    const revisionData: FirebaseOperationDataType = {
+      a: this._userId!,
+      o: operation.toJSON(),
+      t: firebase.firestore.FieldValue.serverTimestamp(), // placeholder for timestamp, it was Number type in RTDB
+    };
+
+    this._doTransaction(revisionId, revisionData, callback);
+  }
+
+  /**
+   * Perform Insert transaction Text Operation into given Revision ID in `history` node of Firebase ref.
+   * @param revisionId - Revision ID.
+   * @param revisionData - Text Operation and metadata in JSON format.
+   * @param callback - Success/Error callback handler.
+   */
+  protected _doTransaction(
+    revisionId: string,
+    revisionData: FirebaseOperationDataType,
+    callback: SendOperationCallbackType
+  ): void {
+    this._firestoreRef!.firestore.runTransaction(async (t) => {
+      t.set(
```

Comment:
In the transaction, we need to check if data already exists in `revisionId`. If it already exists, we need to cancel the transaction.
Two clients can simultaneously push operation with same revisionId (let's say A1). Then if both are allowed to succeed, one client's operation will be overridden by another client. Ref: https://github.com/interviewstreet/firepad-x/blob/236ece5f29b1790e03bd6e3c28cfcc34c91f56ca/src/firebase-adapter.ts#L474-L476

Reply:
Added the check.